src/buffer/out/textBuffer.cpp

```diff
@@ -1928,90 +1928,73 @@ void TextBuffer::_ExpandTextRow(til::inclusive_rect& textRow) const
-
-        // allocate a string buffer
         std::wstring selectionText;
-        std::vector<COLORREF> selectionFgAttr;
-        std::vector<COLORREF> selectionBkAttr;
+        til::small_rle<TextAttribute, uint16_t, 1> attrs;
 
-        // preallocate to avoid reallocs
-        selectionText.reserve(gsl::narrow<size_t>(highlight.Width()) + 2); // + 2 for \r\n if we munged it
-        if (copyTextColor)
-        {
-            selectionFgAttr.reserve(gsl::narrow<size_t>(highlight.Width()) + 2);
-            selectionBkAttr.reserve(gsl::narrow<size_t>(highlight.Width()) + 2);
-        }
+        const auto iRow = selectionRects.at(i).top;
+        const auto& row = GetRowByOffset(iRow);
+        const auto colBegin = gsl::narrow_cast<uint16_t>(selectionRects.at(i).left);
+        auto colEnd = gsl::narrow_cast<uint16_t>(selectionRects.at(i).right + 1); // end column exclusive
 
-        // copy char data into the string buffer, skipping trailing bytes
-        while (it)
+        // We apply formatting to rows if the row was NOT wrapped or formatting
+        // of wrapped rows is allowed.
+        const auto shouldFormatRow = formatWrappedRows || !row.WasWrapForced();
+
+        if (trimTrailingWhitespace && shouldFormatRow)
         {
-            const auto& cell = *it;
+            // update end column to exclude trailing whitespace
 
-            if (cell.DbcsAttr() != DbcsAttribute::Trailing)
+            const auto text = row.GetText();
+            const auto begin = text.rbegin();
+            const auto end = text.rend();
+            auto it = begin;
+            for (; it != end; ++it)
             {
-                const auto chars = cell.Chars();
-                selectionText.append(chars);
-
-                if (copyTextColor)
+                if (*it != L' ')
                 {
-                    const auto cellData = cell.TextAttr();
-                    const auto [CellFgAttr, CellBkAttr] = GetAttributeColors(cellData);
-                    for (size_t j = 0; j < chars.size(); ++j)
-                    {
-                        selectionFgAttr.push_back(CellFgAttr);
-                        selectionBkAttr.push_back(CellBkAttr);
-                    }
+                    break;
                 }
             }
 
-            ++it;
+            const auto textTrimmedLength = gsl::narrow_cast<uint16_t>(end - it);
+            colEnd = std::min(colEnd, textTrimmedLength);
         }
 
-        // We apply formatting to rows if the row was NOT wrapped or formatting of wrapped rows is allowed
-        const auto shouldFormatRow = formatWrappedRows || !GetRowByOffset(iRow).WasWrapForced();
+        // save selected text
+        selectionText = std::wstring{ row.GetText(colBegin, colEnd) };
 
-        if (trimTrailingWhitespace)
+        // retrieve attributes for the text range
+        if (copyAttribute)
         {
-            if (shouldFormatRow)
+            attrs = std::move(row.Attributes().slice(colBegin, colEnd));

```

Comment:
You don't need to move return values, since C++11 and later have RVO - return value optimization. (Also, they're in the so called prvalue category (p = pure) and everything in the more general rvalue category behaves as if it's being moved.)

In fact, this code is a tiny bit dangerous, because if `slice()` was to return a mutable reference, you might steal the contents from the `ROW` instance.

Lastly, I think you can move the `attrs` variable from the other scope in here. Since `slice()` returns a new instance every time there's no way to reuse memory here anyways.